hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -83,27 +87,34 @@ private void invokeOnAllPartitions() {
             Address address = mp.getKey();
             List<Integer> partitions = mp.getValue();
             PartitionIteratingOperation operation = new PartitionIteratingOperation(operationFactory, partitions);
-            Future future = operationService.createInvocationBuilder(serviceName, operation, address)
+
+            InternalCompletableFuture future = operationService.createInvocationBuilder(serviceName, operation, address)
                     .setTryCount(TRY_COUNT)
                     .setTryPauseMillis(TRY_PAUSE_MILLIS)
                     .invoke();
             futures.put(address, future);
         }
     }
 
-    private void awaitCompletion() {
+    private void awaitCompletion() throws Exception {
         NodeEngineImpl nodeEngine = operationService.nodeEngine;
-        for (Map.Entry<Address, Future> response : futures.entrySet()) {
+        for (Map.Entry<Address, InternalCompletableFuture> response : futures.entrySet()) {
             try {
-                Future future = response.getValue();
-                PartitionResponse result = (PartitionResponse) nodeEngine.toObject(future.get());
+                InternalCompletableFuture future = response.getValue();
+                Object o = nodeEngine.toObject(future.get());
+                PartitionResponse result = (PartitionResponse) o;
                 result.addResults(partitionResults);
-            } catch (Throwable t) {
+            } catch (Exception t) {
+                if (!isRetryable(t)) {
```

Comment:
What I can do is:
1: just store all exceptions into the tmp result and move on to the second round.
2: for every failing partition I create an operation (not invoked) and feed it the exception found in the first round. If one of these operation indicates that it can't handle the exception---> throw exception to user. 
3: all operations get invoked.